cirq/linalg/transformations_test.py

```diff
@@ -393,3 +393,169 @@ def test_partial_trace_invalid_inputs():
+#     b = (np.arange(8) + 3) / np.linalg.norm(np.arange(8) + 3)
+#     c = (np.arange(16) + 1) / np.linalg.norm(np.arange(16) + 1)
+#     state = np.kron(np.kron(a, b), c).reshape((2,) * 9)
+#
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [0, 1])), a.reshape(2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [2, 3, 4])), b.reshape(2,2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [5, 6, 7, 8])), c.reshape(2,2,2,2))
+#
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [0, 1, 2, 3, 4])),
+#         np.kron(a, b).reshape(2,2,2,2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [0, 1, 5, 6, 7, 8])),
+#         np.kron(a, c).reshape(2,2,2,2,2,2))
+#     np.testing.assert_almost_equal(
+#         np.abs(cirq.keep_qubits(state, [2, 3, 4, 5, 6, 7, 8])),
+#         np.kron(b, c).reshape(2,2,2,2,2,2,2))
+
+
+def test_keep_qubits_bad_subset():
+    a = np.arange(4) / np.linalg.norm(np.arange(4))
+    b = (np.arange(8) + 3) / np.linalg.norm(np.arange(8) + 3)
+    state = np.kron(a, b).reshape(2, 2, 2, 2, 2)
+    for q1 in range(5):
+        with pytest.raises(ValueError, match='pure'):
+            cirq.keep_qubits(state, [q1])
+    for q1 in range(2):
+        for q2 in range(2, 5):
+            with pytest.raises(ValueError, match='pure'):
+                cirq.keep_qubits(state, [q1, q2])
+    for q3 in range(2, 5):
+        with pytest.raises(ValueError, match='pure'):
+            cirq.keep_qubits(state, [0, 1, q3])
+    for q4 in range(2):
+        with pytest.raises(ValueError, match='pure'):
+            cirq.keep_qubits(state, [2, 3, 4, q4])
+
+
+def test_keep_qubits_non_kron():
+    bell00 = np.array([1, 0, 0, 1]) / np.sqrt(2)
+    for q1 in [0, 1]:
+        with pytest.raises(ValueError, match='pure'):
+            cirq.keep_qubits(bell00.reshape(2,2), [q1])
+
+    plus_x = np.array([1, 1]) / np.sqrt(2)
+    state = np.kron(bell00, plus_x)
+    for q1 in [0, 1]:
+        with pytest.raises(ValueError, match='pure'):
+            cirq.keep_qubits(state.reshape(2,2,2), [q1, 2])
+    np.testing.assert_almost_equal(
+        np.abs(cirq.keep_qubits(state.reshape(2,2,2), [2])), plus_x)
+
+
+def test_keep_qubits_invalid_inputs():
+    with pytest.raises(ValueError, match='normalized'):
+        cirq.keep_qubits(np.arange(16).reshape(2,2,2,2), [1, 2])
+    with pytest.raises(ValueError, match='2, 2'):
+        cirq.keep_qubits(
+            np.arange(16).reshape(2,2,2,2) / np.linalg.norm(np.arange(16)),
+            [1, 2, 2])
+    with pytest.raises(ValueError, match='invalid'):
+        cirq.keep_qubits(np.array([1,0,0,0]).reshape(2,2), [5])
+    with pytest.raises(ValueError, match='invalid'):
+        cirq.keep_qubits(np.array([1,0,0,0]).reshape(2,2), [0, 1, 2])
+
+
+def test_wavefunction_partial_trace_invalid_input():
+    # TODO
```

Comment:
DO

Reply:
I've arranged for `wavefunction_partial_trace` to immediately call `subwavefunction`, and therefore rely on the latter's input validation. Is input validation testing for `wavefunction_partial_trace` still necessary?

Reply:
Yes.

Justifying omitting a test by pointing at the current implementation of the code is very risky business. In a truly ideal world, the tests would be specified completely independent of the implementation so that they cross-check each other. Of course in practice you often use knowledge of the implementation to make the tests more effective, but I prefer to use that as a weapon to make the tests harder rather than a shortcut to make them easier.